core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala

```diff
@@ -196,28 +196,44 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
       case NonFatal(e) => log.warn(e.getMessage, e)
     }
 
+    val hiveTables = Misc.getMemStore.getExternalCatalog.getHiveTables(true).asScala
+
     try {
       // External Tables
-      val hiveTables: java.util.List[ExternalTableMetaData] =
-        Misc.getMemStore.getExternalCatalog.getHiveTables(true)
-      externalTables = hiveTables.asScala.collect {
-        case table if table.tableType.equalsIgnoreCase("EXTERNAL") =>
+      externalTables = hiveTables.collect {
+        case table if table.tableType.equalsIgnoreCase("EXTERNAL") => {
           new SnappyExternalTableStats(table.entityName, table.tableType, table.shortProvider,
             table.externalStore, table.dataSourcePath, table.driverClass)
+        }
       }
     }
     catch {
       case NonFatal(e) => log.warn(e.getMessage, e)
     }
 
-    if(result.flatMap(_.getRegionStats.asScala).size == 0) {
+    if (result.flatMap(_.getRegionStats.asScala).size == 0) {
       // Return last updated tableSizeInfo
       (tableSizeInfo.values.toSeq,
        result.flatMap(_.getIndexStats.asScala),
        externalTables)
     } else {
       // Return updated tableSizeInfo
-      (result.flatMap(_.getRegionStats.asScala),
+      val regionStats = result.flatMap(_.getRegionStats.asScala).map(rs => {
+        hiveTables.map(ht => {
```

Comment:
We can convert hiveTables to a set. Then it will become easier to lookup the current region name in the set. It will avoid iterating the whole list for each table

Reply:
And since we are being too careful here :-) then please change 'hiveTables.map' to 'hiveTables.foreach'
BTW don't we support same column table under different schema? If yes, then won't this
tableName.substring(tableName.indexOf(".") + 1);
mean we are just taking care of table name and not fully qualified names.

Reply:
Adding check on table schema while determining table type.
Changed hiveTables.map to hiveTables.foreach.

Reply:
Also for easier look up, creating and using HashMap built from hiveTables, this will avoid iterating over hiveTables for each and every Table.